// @nktkas/hyperliquid/esm/transport/_polyfills.js
// deno-lint-ignore-file no-explicit-any
/**
 * Runtime shims for APIs missing on some supported platforms, mainly React Native.
 * @module
 */ /** @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise */ export const Promise_ = /* @__PURE__ */ (()=>{
  return {
    /** @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers */ withResolvers: Promise.withResolvers ? ()=>Promise.withResolvers() : ()=>{
      let resolve;
      let reject;
      const promise = new Promise((res, rej)=>(resolve = res, reject = rej));
      return {
        promise,
        resolve,
        reject
      };
    }
  };
})();
/** @see https://developer.mozilla.org/en-US/docs/Web/API/DOMException */ export const DOMException_ = /* @__PURE__ */ (()=>{
  return globalThis.DOMException || class DOMException extends Error {
    constructor(message = "", name = "Error"){
      super(message);
      this.name = name;
    }
  };
})();
/** @see https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent */ export const CustomEvent_ = /* @__PURE__ */ (()=>{
  return globalThis.CustomEvent || class CustomEvent extends Event {
    detail;
    constructor(type, eventInitDict){
      super(type, eventInitDict);
      this.detail = eventInitDict?.detail ?? null;
    }
    initCustomEvent() {}
  };
})();
